File: fe/mobile/lib/models/payment_card.dart
class PaymentCard {
  const PaymentCard({
    required this.id,
    required this.holderName,
    required this.network,
    required this.last4,
    required this.expiryMonth,
    required this.expiryYear,
  });

  final String id;
  final String holderName;
  final String network;
  final String last4;
  final int expiryMonth;
  final int expiryYear;

  String get maskedNumber => '**** **** **** $last4';
  String get expiryLabel =>
      '${expiryMonth.toString().padLeft(2, '0')}/$expiryYear';

  factory PaymentCard.fromJson(Map<String, dynamic> json) {
    return PaymentCard(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      holderName: json['holderName']?.toString() ?? '',
      network: json['network']?.toString() ?? '',
      last4: json['last4']?.toString() ?? '',
      expiryMonth: (json['expiryMonth'] is num)
          ? (json['expiryMonth'] as num).toInt()
          : 0,
      expiryYear: (json['expiryYear'] is num)
          ? (json['expiryYear'] as num).toInt()
          : 0,
    );
  }
}

File: be/src/models/payment_card.ts
import { CardNetwork } from "@dtos/request/card.request";
import mongoose, { Document, Model } from "mongoose";

export interface IPaymentCard extends Document {
  userId: mongoose.Types.ObjectId;
  holderName: string;
  network: CardNetwork;
  last4: string;
  expiryMonth: number;
  expiryYear: number;
  cardHash: string;
  cvvHash: string;
  fingerprintHash: string;
  isActive: boolean;
  createdAt: Date;
  updatedAt: Date;
}

class PaymentCardModel {
  private model: Model<IPaymentCard>;

  constructor() {
    this.model = mongoose.model<IPaymentCard>(
      "PaymentCard",
      new mongoose.Schema(
        {
          userId: {
            type: mongoose.Schema.Types.ObjectId,
            ref: "User",
            required: true,
            index: true,
          },
          holderName: { type: String, required: true, trim: true },
          network: {
            type: String,
            enum: ["VISA", "MASTERCARD", "UNIONPAY"],
            required: true,
          },
          last4: { type: String, required: true },
          expiryMonth: { type: Number, required: true },
          expiryYear: { type: Number, required: true },
          cardHash: { type: String, required: true },
          cvvHash: { type: String, required: true },
          fingerprintHash: { type: String, required: true },
          isActive: { type: Boolean, default: true, index: true },
        },
        { timestamps: true }
      )
    );
  }

  getModel(): Model<IPaymentCard> {
    return this.model;
  }
}

export default new PaymentCardModel();
